File: python/lattigo/lattigo/rlwe.py
"""RLWE types: keys, ciphertext, plaintext, encryptor, decryptor, key generator.

Wraps raw Lattigo rlwe types via the bridge FFI.
"""

from __future__ import annotations

from . import ffi
from .ckks import Parameters
from .gohandle import GoHandle

# =========================================================================
# Key types
# =========================================================================


class SecretKey:
    """Wraps Lattigo's rlwe.SecretKey."""

    def __init__(self, handle: GoHandle):
        self._handle = handle

    def marshal_binary(self) -> bytes:
        return ffi.secret_key_marshal(self._handle)

    @classmethod
    def unmarshal_binary(cls, data: bytes) -> SecretKey:
        return cls(ffi.secret_key_unmarshal(data))

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> SecretKey:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


class PublicKey:
    """Wraps Lattigo's rlwe.PublicKey."""

    def __init__(self, handle: GoHandle):
        self._handle = handle

    def marshal_binary(self) -> bytes:
        return ffi.public_key_marshal(self._handle)

    @classmethod
    def unmarshal_binary(cls, data: bytes) -> PublicKey:
        return cls(ffi.public_key_unmarshal(data))

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> PublicKey:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


class RelinearizationKey:
    """Wraps Lattigo's rlwe.RelinearizationKey."""

    def __init__(self, handle: GoHandle):
        self._handle = handle

    def marshal_binary(self) -> bytes:
        return ffi.relin_key_marshal(self._handle)

    @classmethod
    def unmarshal_binary(cls, data: bytes) -> RelinearizationKey:
        return cls(ffi.relin_key_unmarshal(data))

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> RelinearizationKey:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


class GaloisKey:
    """Wraps Lattigo's rlwe.GaloisKey."""

    def __init__(self, handle: GoHandle):
        self._handle = handle

    def marshal_binary(self) -> bytes:
        return ffi.galois_key_marshal(self._handle)

    @classmethod
    def unmarshal_binary(cls, data: bytes) -> GaloisKey:
        return cls(ffi.galois_key_unmarshal(data))

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> GaloisKey:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


# =========================================================================
# Ciphertext / Plaintext
# =========================================================================


class Ciphertext:
    """Wraps a raw Lattigo rlwe.Ciphertext (single ciphertext, no shape)."""

    def __init__(self, handle: GoHandle):
        self._handle = handle

    def level(self) -> int:
        return ffi.rlwe_ciphertext_level(self._handle)

    def marshal_binary(self) -> bytes:
        return ffi.rlwe_ciphertext_marshal(self._handle)

    @classmethod
    def unmarshal_binary(cls, data: bytes) -> Ciphertext:
        return cls(ffi.rlwe_ciphertext_unmarshal(data))

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> Ciphertext:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


class Plaintext:
    """Wraps a raw Lattigo rlwe.Plaintext."""

    def __init__(self, handle: GoHandle):
        self._handle = handle

    def level(self) -> int:
        return ffi.rlwe_plaintext_level(self._handle)

    def marshal_binary(self) -> bytes:
        return ffi.rlwe_plaintext_marshal(self._handle)

    @classmethod
    def unmarshal_binary(cls, data: bytes) -> Plaintext:
        return cls(ffi.rlwe_plaintext_unmarshal(data))

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> Plaintext:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


# =========================================================================
# KeyGenerator
# =========================================================================


class KeyGenerator:
    """Wraps Lattigo's rlwe.KeyGenerator."""

    def __init__(self, params: Parameters):
        self._handle = ffi.new_key_generator(params._h)

    @classmethod
    def _from_handle(cls, handle: GoHandle) -> KeyGenerator:
        """Wrap an existing Go handle (internal use)."""
        obj = object.__new__(cls)
        obj._handle = handle
        return obj

    def gen_secret_key(self) -> SecretKey:
        return SecretKey(ffi.keygen_gen_secret_key(self._handle))

    def gen_public_key(self, sk: SecretKey) -> PublicKey:
        return PublicKey(ffi.keygen_gen_public_key(self._handle, sk._handle))

    def gen_relin_key(self, sk: SecretKey) -> RelinearizationKey:
        return RelinearizationKey(ffi.keygen_gen_relin_key(self._handle, sk._handle))

    def gen_galois_key(self, sk: SecretKey, galois_element: int) -> GaloisKey:
        return GaloisKey(ffi.keygen_gen_galois_key(self._handle, sk._handle, galois_element))

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> KeyGenerator:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


# =========================================================================
# Encryptor / Decryptor
# =========================================================================


class Encryptor:
    """Wraps Lattigo's rlwe.Encryptor (public-key encryption)."""

    def __init__(self, params: Parameters, pk: PublicKey):
        self._handle = ffi.new_encryptor(params._h, pk._handle)

    @classmethod
    def _from_handle(cls, handle: GoHandle) -> Encryptor:
        """Wrap an existing Go handle (internal use)."""
        obj = object.__new__(cls)
        obj._handle = handle
        return obj

    def encrypt_new(self, pt: Plaintext) -> Ciphertext:
        ct_h = ffi.encryptor_encrypt_new(self._handle, pt._handle)
        return Ciphertext(ct_h)

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> Encryptor:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


class Decryptor:
    """Wraps Lattigo's rlwe.Decryptor."""

    def __init__(self, params: Parameters, sk: SecretKey):
        self._handle = ffi.new_decryptor(params._h, sk._handle)

    @classmethod
    def _from_handle(cls, handle: GoHandle) -> Decryptor:
        """Wrap an existing Go handle (internal use)."""
        obj = object.__new__(cls)
        obj._handle = handle
        return obj

    def decrypt_new(self, ct: Ciphertext) -> Plaintext:
        pt_h = ffi.decryptor_decrypt_new(self._handle, ct._handle)
        return Plaintext(pt_h)

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> Decryptor:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


# =========================================================================
# MemEvaluationKeySet
# =========================================================================


class MemEvaluationKeySet:
    """Wraps Lattigo's rlwe.MemEvaluationKeySet."""

    def __init__(
        self,
        rlk: RelinearizationKey | None = None,
        galois_keys: list[GaloisKey] | None = None,
    ):
        self._handle = ffi.new_mem_eval_key_set(
            rlk._handle if rlk else None,
            [gk._handle for gk in (galois_keys or [])],
        )

    @classmethod
    def _from_handle(cls, handle: GoHandle) -> MemEvaluationKeySet:
        """Wrap an existing Go handle (internal use)."""
        obj = object.__new__(cls)
        obj._handle = handle
        return obj

    def marshal_binary(self) -> bytes:
        return ffi.mem_eval_key_set_marshal(self._handle)

    @classmethod
    def unmarshal_binary(cls, data: bytes) -> MemEvaluationKeySet:
        return cls._from_handle(ffi.mem_eval_key_set_unmarshal(data))

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> MemEvaluationKeySet:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


# =========================================================================
# Bootstrap
# =========================================================================


class BootstrapEvalKeys:
    """Wraps Lattigo's bootstrapping.EvaluationKeys (serializable bootstrap keys)."""

    def __init__(self, handle: GoHandle):
        self._handle = handle

    def marshal_binary(self) -> bytes:
        return ffi.bootstrap_eval_keys_marshal(self._handle)

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> BootstrapEvalKeys:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


class BootstrapParams:
    """Wraps Lattigo's bootstrapping parameters.

    Used to generate the evaluation keys required for bootstrapping
    (ciphertext refresh) during FHE inference.

    Example::

        btp = BootstrapParams(params, logn=14, logp=[61]*6, h=192, log_slots=7)
        evk, btp_keys = btp.gen_eval_keys(sk)
        btp_keys_bytes = btp_keys.marshal_binary()
    """

    def __init__(
        self,
        params: Parameters,
        *,
        logn: int = 0,
        logp: list[int] | None = None,
        h: int = 0,
        log_slots: int = 0,
    ):
        self._handle = ffi.new_bootstrap_params(
            params._h, logn=logn, logp=logp, h=h, log_slots=log_slots,
        )

    def gen_eval_keys(
        self, sk: SecretKey,
    ) -> tuple[MemEvaluationKeySet, BootstrapEvalKeys]:
        """Generate bootstrap evaluation keys.

        Returns:
            A tuple of (evk, btp_keys) where:
            - evk: MemEvaluationKeySet for the evaluator
            - btp_keys: BootstrapEvalKeys for serialization
        """
        evk_h, btp_evk_h = ffi.bootstrap_params_gen_eval_keys(
            self._handle, sk._handle,
        )
        return MemEvaluationKeySet._from_handle(evk_h), BootstrapEvalKeys(btp_evk_h)

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> BootstrapParams:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass

File: js/lattigo/src/rlwe.ts
import type { Handle } from "./types.js";
import { isError } from "./types.js";
import { getBridge, registry } from "./bridge.js";
import type { CKKSParameters } from "./ckks.js";

// =========================================================================
// Key types
// =========================================================================

/** Wraps Lattigo's rlwe.SecretKey. */
export class SecretKey {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  marshalBinary(): Uint8Array {
    const result = getBridge().secretKeyMarshal(this._handle);
    if (isError(result)) throw new Error(`SecretKey marshal: ${result.error}`);
    return result;
  }

  static unmarshalBinary(bytes: Uint8Array): SecretKey {
    const result = getBridge().secretKeyUnmarshal(bytes);
    if (isError(result))
      throw new Error(`SecretKey unmarshal: ${result.error}`);
    return new SecretKey(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

/** Wraps Lattigo's rlwe.PublicKey. */
export class PublicKey {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  marshalBinary(): Uint8Array {
    const result = getBridge().publicKeyMarshal(this._handle);
    if (isError(result)) throw new Error(`PublicKey marshal: ${result.error}`);
    return result;
  }

  static unmarshalBinary(bytes: Uint8Array): PublicKey {
    const result = getBridge().publicKeyUnmarshal(bytes);
    if (isError(result))
      throw new Error(`PublicKey unmarshal: ${result.error}`);
    return new PublicKey(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

/** Wraps Lattigo's rlwe.RelinearizationKey. */
export class RelinearizationKey {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  marshalBinary(): Uint8Array {
    const result = getBridge().relinKeyMarshal(this._handle);
    if (isError(result))
      throw new Error(`RelinearizationKey marshal: ${result.error}`);
    return result;
  }

  static unmarshalBinary(bytes: Uint8Array): RelinearizationKey {
    const result = getBridge().relinKeyUnmarshal(bytes);
    if (isError(result))
      throw new Error(`RelinearizationKey unmarshal: ${result.error}`);
    return new RelinearizationKey(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

/** Wraps Lattigo's rlwe.GaloisKey. */
export class GaloisKey {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  marshalBinary(): Uint8Array {
    const result = getBridge().galoisKeyMarshal(this._handle);
    if (isError(result)) throw new Error(`GaloisKey marshal: ${result.error}`);
    return result;
  }

  static unmarshalBinary(bytes: Uint8Array): GaloisKey {
    const result = getBridge().galoisKeyUnmarshal(bytes);
    if (isError(result))
      throw new Error(`GaloisKey unmarshal: ${result.error}`);
    return new GaloisKey(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

// =========================================================================
// Ciphertext / Plaintext
// =========================================================================

/** Wraps a raw Lattigo rlwe.Ciphertext. */
export class Ciphertext {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  level(): number {
    const result = getBridge().ciphertextLevel(this._handle);
    if (isError(result))
      throw new Error(`Ciphertext level: ${result.error}`);
    return result;
  }

  marshalBinary(): Uint8Array {
    const result = getBridge().ciphertextMarshal(this._handle);
    if (isError(result))
      throw new Error(`Ciphertext marshal: ${result.error}`);
    return result;
  }

  static unmarshalBinary(bytes: Uint8Array): Ciphertext {
    const result = getBridge().ciphertextUnmarshal(bytes);
    if (isError(result))
      throw new Error(`Ciphertext unmarshal: ${result.error}`);
    return new Ciphertext(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

/** Wraps a raw Lattigo rlwe.Plaintext. */
export class Plaintext {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  level(): number {
    const result = getBridge().plaintextLevel(this._handle);
    if (isError(result))
      throw new Error(`Plaintext level: ${result.error}`);
    return result;
  }

  marshalBinary(): Uint8Array {
    const result = getBridge().plaintextMarshal(this._handle);
    if (isError(result))
      throw new Error(`Plaintext marshal: ${result.error}`);
    return result;
  }

  static unmarshalBinary(bytes: Uint8Array): Plaintext {
    const result = getBridge().plaintextUnmarshal(bytes);
    if (isError(result))
      throw new Error(`Plaintext unmarshal: ${result.error}`);
    return new Plaintext(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

// =========================================================================
// KeyGenerator
// =========================================================================

/** Wraps Lattigo's rlwe.KeyGenerator. */
export class KeyGenerator {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  static new(params: CKKSParameters): KeyGenerator {
    const result = getBridge().newKeyGenerator(params.handle);
    if (isError(result))
      throw new Error(`KeyGenerator create: ${result.error}`);
    return new KeyGenerator(result.handle);
  }

  genSecretKey(): SecretKey {
    const result = getBridge().keyGenGenSecretKey(this._handle);
    if (isError(result))
      throw new Error(`genSecretKey: ${result.error}`);
    return new SecretKey(result.handle);
  }

  genPublicKey(sk: SecretKey): PublicKey {
    const result = getBridge().keyGenGenPublicKey(this._handle, sk.handle);
    if (isError(result))
      throw new Error(`genPublicKey: ${result.error}`);
    return new PublicKey(result.handle);
  }

  genRelinKey(sk: SecretKey): RelinearizationKey {
    const result = getBridge().keyGenGenRelinKey(this._handle, sk.handle);
    if (isError(result))
      throw new Error(`genRelinKey: ${result.error}`);
    return new RelinearizationKey(result.handle);
  }

  genGaloisKey(sk: SecretKey, galoisElement: number): GaloisKey {
    const result = getBridge().keyGenGenGaloisKey(
      this._handle,
      sk.handle,
      galoisElement,
    );
    if (isError(result))
      throw new Error(`genGaloisKey: ${result.error}`);
    return new GaloisKey(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

// =========================================================================
// Encryptor / Decryptor
// =========================================================================

/** Wraps Lattigo's rlwe.Encryptor (public-key encryption). */
export class Encryptor {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  static new(params: CKKSParameters, pk: PublicKey): Encryptor {
    const result = getBridge().newEncryptor(params.handle, pk.handle);
    if (isError(result))
      throw new Error(`Encryptor create: ${result.error}`);
    return new Encryptor(result.handle);
  }

  encryptNew(pt: Plaintext): Ciphertext {
    const result = getBridge().encryptorEncryptNew(this._handle, pt.handle);
    if (isError(result))
      throw new Error(`encryptNew: ${result.error}`);
    return new Ciphertext(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

/** Wraps Lattigo's rlwe.Decryptor. */
export class Decryptor {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  static new(params: CKKSParameters, sk: SecretKey): Decryptor {
    const result = getBridge().newDecryptor(params.handle, sk.handle);
    if (isError(result))
      throw new Error(`Decryptor create: ${result.error}`);
    return new Decryptor(result.handle);
  }

  decryptNew(ct: Ciphertext): Plaintext {
    const result = getBridge().decryptorDecryptNew(this._handle, ct.handle);
    if (isError(result))
      throw new Error(`decryptNew: ${result.error}`);
    return new Plaintext(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

// =========================================================================
// MemEvaluationKeySet
// =========================================================================

/** Wraps Lattigo's rlwe.MemEvaluationKeySet. */
export class MemEvaluationKeySet {
  private _handle: Handle;

  constructor(handle: Handle) {
    this._handle = handle;
    registry.register(this, handle, this);
  }

  get handle(): Handle {
    return this._handle;
  }

  /**
   * Create a MemEvaluationKeySet from optional RLK + Galois keys.
   * Pass null for rlk if no relinearization key is needed.
   */
  static new(
    rlk: RelinearizationKey | null,
    galoisKeys: GaloisKey[] = [],
  ): MemEvaluationKeySet {
    const result = getBridge().newMemEvalKeySet(
      rlk?.handle ?? null,
      galoisKeys.map((gk) => gk.handle),
    );
    if (isError(result))
      throw new Error(`MemEvaluationKeySet create: ${result.error}`);
    return new MemEvaluationKeySet(result.handle);
  }

  marshalBinary(): Uint8Array {
    const result = getBridge().memEvalKeySetMarshal(this._handle);
    if (isError(result))
      throw new Error(`MemEvaluationKeySet marshal: ${result.error}`);
    return result;
  }

  static unmarshalBinary(bytes: Uint8Array): MemEvaluationKeySet {
    const result = getBridge().memEvalKeySetUnmarshal(bytes);
    if (isError(result))
      throw new Error(`MemEvaluationKeySet unmarshal: ${result.error}`);
    return new MemEvaluationKeySet(result.handle);
  }

  free(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}
